### PR TITLE
hooks for plugins to override Global.analyzer

### PR DESCRIPTION
**Pull request by xeno-by**
As discussed a couple weeks ago, submitting this to formalize the hacks
that I used in macro paradise for 2.10, so that it continues working
after possible refactorings in 2.11.

One of the strategies to implement this would be to make a list of
functions overridden by the plugin and then to reuse the analyzer
plugins architecture to introduce them as hooks. However:
1) The list is quite big and disparate (Namers.enterSym,
Namers.ensureCompanionObject, Typers.reallyExists, Typers.typed1,
Typers.typedTemplate, Typers.typedBlock, Typers.typedClassDef).
2) The list doesn't correspond to the spirit of analyzer plugins that
modify return types of namer/typer functions rather than override them.
3) There's no guarantee that later on paradise won't need to override
additional functionality of namer/typer, requiring new hacks.

Another strategy lets plugins suggest their own analyzers to the compiler.
It enables augmenting namer/typer with arbitrary functionality, providing
ultimate flexibility. Shortcomings:
1) If multiple plugins want to install their own analyzers, it won't work
(but such situations can be detected and meaningful errors can be reported).
One can argue though, that even if we provided finer-grained API for
such plugins, it wouldn't work anyway, because e.g. it's hard to imagine
several plugins robustly applying their modifications to Namers.enterSym.

None of the outlined two strategies is ideal, however, the second one
makes more sense to me, because shortcoming #3 of the first one seems to
defeat the idea of introducing the hooks into the compiler in the first place.
Therefore this patch implements the second strategy.

### CODE REVIEW

**Comment by xeno-by** [comment]
review @adriaanm @gkossakowski @retronym @som-snytt

File: src/partest/scala/tools/partest/nest/FileManager.scala [hunk 1]
```diff
@@ -158,7 +158,7 @@ trait FileManager extends FileUtil {
     val xprefix          = "-Xplugin:"
     val (xplugs, others) = args partition (_ startsWith xprefix)
     val Xplugin          = if (xplugs.isEmpty) Nil else List(xprefix +
-      (xplugs map (_ stripPrefix xprefix) flatMap (_ split pathSeparator) map absolutize mkString pathSeparator)
+      (xplugs map (_ stripPrefix xprefix) flatMap (_ split ",") map absolutize mkString ",")
```
Comment (gkossakowski) [inline src/partest/scala/tools/partest/nest/FileManager.scala:161]: Why that change?

Comment (xeno-by) [inline src/partest/scala/tools/partest/nest/FileManager.scala:161]: I believe that old code was incorrect. Entries in the `-Xplugin:...` list are separated with commas, not colons or semicolons.

Comment (som-snytt) [inline src/partest/scala/tools/partest/nest/FileManager.scala:161]: ! Indeed, this ancient bit of code has been refactored or rescued after the paulp refactor, and I never noticed that.  It's not very useful to specify two locations right now, which is a shortcoming.

**Comment by xeno-by** [comment]
Looks like the failures are part spurious and part not.

[neg/analyzerReplacingPlugins-a](https://github.com/scala/scala/pull/2846/files#L5R1): looks spurious, since a scala-checkin-manual job didn't have this problem: https://scala-webapps.epfl.ch/jenkins/view/All/job/scala-checkin-manual/995/consoleText. What can we do about that?

[run/t6745-2.scala](https://github.com/xeno-by/scala/blob/topic/paradise-hooks/test/files/run/t6745-2.scala) and [run/t4426.scala](https://github.com/xeno-by/scala/blob/topic/paradise-hooks/test/files/run/t4426.scala) fail because of inliner warnings saying something like:

```
t4426.scala:13: warning: Could not inline required method assert because bytecode unavailable.
        assert(dummy ne null, "Dummy not initialized")
              ^
```

Both of the tests that fail because of the inliner somehow touch the things changed in this commit (e.g. `global.analyzer`), but I fail to understand how that could prevent the inliner from inlining `Predef.assert`. Any ideas?

**Comment by gkossakowski** [comment]
I'll look into this PR on Thursday.

**Comment by gkossakowski** [comment]
@xeno-by: thanks for your analysis. However, I don't agree with your conclusions.

You said:

> 3) There's no guarantee that later on paradise won't need to override additional functionality of namer/typer, requiring new hacks.

That's true but that's the point of having small surface area in your api: you expose only minimal API that is needed by their clients to implement what they want. If it turns out that the API is too small for given use case you expand it over time.

The alternative you propose is to expose the whole `Analyzer` as an API. The problem with that is this:

```
scala> typeOf[scala.tools.nsc.typechecker.Analyzer]
res1: $r.intp.global.Type = scala.tools.nsc.typechecker.Analyzer

scala> res1.members.size
res2: Int = 372
```

We make 372 members as potential targets for API calls. This way you make it very difficult to change anything in compiler internals because they are exposed to plugins.

I really think we should start with minimal interface and evolve it according to the needs of API clients. Note that compiler API is not restricted by binary compatibility the same as scala library and reflect are for now.

@retronym: I'd love to hear your thoughts on this too.

**Comment by gkossakowski** [comment]
Also, this doesn't merge anymore so I'm closing it.

I think we should discuss this on internals before next PR is issued.

**Comment by xeno-by** [comment]
@gkossakowski https://groups.google.com/forum/#!topic/scala-internals/q8y9k4EoBNg

**Pull request closed without merging**